### PR TITLE
migrate to Celery 4

### PR DESCRIPTION
**Pull request by domdinicola**

### CODE REVIEW

File: EquiTrack/EquiTrack/settings/base.py [hunk 1]
```diff
@@ -300,20 +300,18 @@ def str2bool(value):
...
 }
 
-# django-celery: https://github.com/celery/django-celery
-djcelery.setup_loader()
-
 # celery: http://docs.celeryproject.org/en/3.1/configuration.html
-BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
+CELERY_BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
 BROKER_VISIBILITY_VAR = os.environ.get('CELERY_VISIBILITY_TIMEOUT', 1800)  # in seconds
 BROKER_TRANSPORT_OPTIONS = {'visibility_timeout': int(BROKER_VISIBILITY_VAR)}
```
Comment (vkurup) [inline EquiTrack/EquiTrack/settings/base.py]: When I ran `celery upgrade settings EquiTrack/settings/base.py --django`, it recommended that I change this to `CELERY_BROKER_TRANSPORT_OPTIONS`

File: EquiTrack/EquiTrack/settings/base.py [hunk 2]
```diff
@@ -300,20 +300,18 @@ def str2bool(value):
...
-CELERY_ACKS_LATE = True
+CELERY_TASK_ALWAYS_EAGER = CELERY_ALWAYS_EAGER = False
+CELERY_TASK_ACKS_LATE = CELERY_ACKS_LATE = True
 CELERY_TASK_PUBLISH_RETRY = True
-CELERY_DISABLE_RATE_LIMITS = False
+CELERY_WORKER_DISABLE_RATE_LIMITS = CELERY_DISABLE_RATE_LIMITS = False
+
 # By default we will ignore result
 # If you want to see results and try out tasks interactively, change it to False
 # Or change this setting on tasks level
```
Comment (vkurup) [inline EquiTrack/EquiTrack/settings/base.py]: celery's upgrade tool also recommended changes in the settings below:

```
 # By default we will ignore result
 # If you want to see results and try out tasks interactively, change it to False
 # Or change this setting on tasks level
-CELERY_IGNORE_RESULT = True
+CELERY_TASK_IGNORE_RESULT = True
 CELERY_SEND_TASK_ERROR_EMAILS = False
-CELERY_TASK_RESULT_EXPIRES = 600
-CELERYD_PREFETCH_MULTIPLIER = 1
+CELERY_RESULT_EXPIRES = 600
+CELERY_WORKER_PREFETCH_MULTIPLIER = 1
```

File: EquiTrack/EquiTrack/settings/base.py [hunk 3]
```diff
@@ -300,20 +300,18 @@ def str2bool(value):
...
 BROKER_TRANSPORT_OPTIONS = {'visibility_timeout': int(BROKER_VISIBILITY_VAR)}
-CELERY_RESULT_BACKEND = 'djcelery.backends.database:DatabaseBackend'
-CELERYBEAT_SCHEDULER = 'djcelery.schedulers.DatabaseScheduler'
+CELERY_RESULT_BACKEND = 'django_celery_results.backends.database:DatabaseBackend'
+CELERY_BEAT_SCHEDULER = 'django_celery_beat.schedulers.DatabaseScheduler'
 # Sensible settings for celery
-CELERY_ALWAYS_EAGER = False
-CELERY_ACKS_LATE = True
+CELERY_TASK_ALWAYS_EAGER = CELERY_ALWAYS_EAGER = False
+CELERY_TASK_ACKS_LATE = CELERY_ACKS_LATE = True
```
Comment (vkurup) [inline EquiTrack/EquiTrack/settings/base.py]: Do we need to keep the old names?

Comment (domdinicola) [reply EquiTrack/EquiTrack/settings/base.py]: I guess we can get rid of those

File: EquiTrack/EquiTrack/settings/base.py (hunk 1 above)
Comment (domdinicola) [reply EquiTrack/EquiTrack/settings/base.py]: perfect

File: EquiTrack/EquiTrack/settings/base.py (hunk 2 above)
Comment (domdinicola) [reply EquiTrack/EquiTrack/settings/base.py]: ok

**Comment by domdinicola** [comment]
I'll create another PR in order to move before to 3.1.25 (should not be necessary, but we don't need to rush to upgrade)

**Comment by domdinicola** [comment]
do you have any idea about the reason blocking the build?
django_celery_beat and django_celery_results are new apps with migrations

**Comment by vkurup** [comment]
Thanks for these changes, looks good!

I think there is a problem with django-celery-beat. I didn't notice it when running locally, but I re-ran the build on CircleCI with SSH enabled and when I ran `makemigrations`, I got this:

```
Migrations for 'django_celery_beat':
  0005_auto_20171115_2343.py:
    - Alter field event on solarschedule
```

I think it's due to this issue: https://github.com/celery/django-celery-beat/issues/91

There are a couple PRs at that repo which should fix the issue but neither has been merged yet...

**Comment by domdinicola** [comment]
@vkurup great! Let's wait for a fixing version then

**Comment by vkurup** [comment]
Sounds good @domdinicola!

**Comment by dpoirier** [comment]
When work on this is continued, we can simplify the retry code in vision/tasks.py by using the autoretry_for argument to the task decorator.

File: EquiTrack/EquiTrack/settings/base.py [hunk 4]
```diff
@@ -298,27 +298,25 @@ def str2bool(value):
     }
 }
 
-# django-celery: https://github.com/celery/django-celery
-djcelery.setup_loader()
-
 # celery: http://docs.celeryproject.org/en/3.1/configuration.html
-BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
+CELERY_BROKER_URL = os.environ.get('REDIS_URL', 'redis://localhost:6379/0')
 BROKER_VISIBILITY_VAR = os.environ.get('CELERY_VISIBILITY_TIMEOUT', 1800)  # in seconds
```
Comment (robertavram) [inline EquiTrack/EquiTrack/settings/base.py]: why did you leave out the visibility_var?

Comment (domdinicola) [reply EquiTrack/EquiTrack/settings/base.py]: this is not a default celery setting, but I guess is better to add the namespace to it anyway

**Comment by robertavram** [comment]
@domdinicola tests are failing.

**Comment by domdinicola** [comment]
@robertavram is passing now

File: EquiTrack/EquiTrack/settings/base.py [hunk 5]
```diff
@@ -155,7 +154,8 @@ def str2bool(value):
     'smart_selects',
     'gunicorn',
     'post_office',
-    'djcelery',
```
Comment (ntrncic) [inline EquiTrack/EquiTrack/settings/base.py]: @domdinicola why did we remove django-celery? In order to run celerycam we need djcelery.

**Comment by domdinicola** [comment]
from celery 4 you don't need anymore djcelery:
https://github.com/celery/django-celery
http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html

**Pull request merged**